app/controllers/CompilerService.scala

```diff
@@ -203,14 +205,23 @@ object CompilerService extends Controller {
   private def extractModelString(argMap: Map[String, String], missingMsg: String): ValidationNel[String, String] =
     (argMap get modelKey).fold(missingMsg.failureNel[String])(_.successNel[String])
 
-  private def fetchURL(url: String): ValidationNel[String, String] = {
-    Try(new URL(url)) map {
-      wellFormedURL => usingSource(_.fromURL(wellFormedURL))(_.mkString).successNel
-    } recover {
-      case _: MalformedURLException => s"'$url' is an invalid URL.".failureNel
-    } getOrElse {
-      s"An unknown error occurred while processing the URL '$url'. Make sure the url is publicly accessible.".failureNel
-    }
+  private def fetchURL(url: String, hostUri: String): ValidationNel[String, String] = {
+    val locallyHostedRegex = s"^https?://$hostUri([A-Za-z0-9%/]+\\.nlogo)$$".r
+    locallyHostedRegex.findFirstIn(url).flatMap {
+      case locallyHostedRegex(file) =>
+        Assets.resourceNameAt("public", file).map(path =>
+          usingSource(_.fromFile(path))(_.mkString).successNel
```

Comment:
Bah, using `Assets` still leaves you reading a file from the disk unless you want to use `Assets.at` and bubble an `Action` up the call stack (which you probably don't want to do).

The point of my earlier suggestion was to leverage the fact that Play already makes asset loading fast by putting the assets on the classpath.  I guess I would suggest grabbing the file from the classpath directly with `Play.resource(s"/public/$file")` (which returns you its classpath URL, which I would expect you to be able to read from speedily with `usingSource`).